File: go/tree/rbTree/rbtree.go
package rbtree

import (
	"fmt"
)

type NodeColor int

const (
	Red NodeColor = iota
	Black
)

type RBNode struct {
	Value  int
	Left   *RBNode
	Parent *RBNode
	Right  *RBNode
	Color  NodeColor
}

type RBTree struct {
	root *RBNode
	nil  *RBNode
}

func NewRBTree() *RBTree {
	nilNode := &RBNode{
		Value: -1,
		Color: Black,
	}
	return &RBTree{
		nil: nilNode,
	}
}

func (t *RBTree) Insert(value int) {
	node := &RBNode{
		Value:  value,
		Left:   t.nil,
		Parent: t.nil,
		Right:  t.nil,
		Color:  Red,
	}
	t.root = t.insert(t.root, node, t.nil)
	t.insertFixup(node)
}

func (t *RBTree) insert(node, newNode, parent *RBNode) *RBNode {
	if node == t.nil || node == nil {
		newNode.Parent = parent
		return newNode
	}
	if node.Value < newNode.Value {
		node.Right = t.insert(node.Right, newNode, node)
	} else if node.Value > newNode.Value {
		node.Left = t.insert(node.Left, newNode, node)
	}
	return node
}

// 左旋转
func (t *RBTree) rotateLeft(node *RBNode) {
	right := node.Right
	node.Right = right.Left

	if right.Left != t.nil {
		right.Left.Parent = node
	}
	right.Parent = node.Parent
	if node.Parent == t.nil {
		t.root = right
	} else if node == node.Parent.Left {
		node.Parent.Left = right
	} else {
		node.Parent.Right = right
	}

	right.Left = node
	node.Parent = right
}

// 右旋

func (t *RBTree) rotateRight(node *RBNode) {
	left := node.Left
	node.Left = left.Right

	if left.Right != t.nil {
		left.Right.Parent = node
	}
	left.Parent = node.Parent
	if node.Parent == t.nil {
		t.root = left
	} else if node == node.Parent.Left {
		node.Parent.Left = left
	} else {
		node.Parent.Right = left
	}

	left.Right = node
	node.Parent = left
}

func (t *RBTree) insertFixup(node *RBNode) {
	// 当node->parent的颜色为黑色时或者node是根节点时（因为node为根节点是只有可能违反第二条性质，但在代码最后有将根节点标记为黑色）终止循环

	for node != t.root && node.Parent.Color == Red {
		// node的父节点是其父节点的左孩子
		if node.Parent == node.Parent.Parent.Left {
			// node的叔叔节点
			uncle := node.Parent.Parent.Right
			// 如果叔叔节点颜色为红色
			if uncle != nil && uncle.Color == Red {
				// 将其父节点和叔叔节点的颜色都变为黑色
				node.Parent.Color = Black
				uncle.Color = Black
				// 将其父节点的父节点的颜色变为红色
				node.Parent.Parent.Color = Red
				// 将node指向其父节点的父节点
				node = node.Parent.Parent
			} else {
				// 如果node是其父节点的右孩子
				if node == node.Parent.Right {

					node = node.Parent
					t.rotateLeft(node)
				}

				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				t.rotateRight(node.Parent.Parent)
			}
		} else {
			// node的父节点是其父节点的右孩子
			// node的叔叔节点
			uncle := node.Parent.Parent.Left
			// 如果叔叔节点颜色为红色
			if uncle != nil && uncle.Color == Red {
				// 将其父节点和叔叔节点的颜色都变为黑色
				node.Parent.Color = Black
				uncle.Color = Black
				// 将其父节点的父节点的颜色变为红色
				node.Parent.Parent.Color = Red
				// 将node指向其父节点的父节点
				node = node.Parent.Parent
			} else {
				// 如果node是其父节点的左孩子
				if node == node.Parent.Left {
					node = node.Parent
					t.rotateRight(node)
				}
				node.Parent.Color = Black
				node.Parent.Parent.Color = Red
				t.rotateLeft(node.Parent.Parent)
			}
		}
	}
	t.root.Color = Black
}

func (t *RBTree) Search(value int) *RBNode {
	return t.search(t.root, value)
}
func (t *RBTree) search(node *RBNode, value int) *RBNode {
	if node == t.nil || node == nil {
		return nil
	}
	if node.Value == value {
		return node
	}
	if node.Value < value {
		return t.search(node.Right, value)
	}
	return t.search(node.Left, value)
}

func (t *RBTree) Delete(value int) {
	node := t.Search(value)
	if node == nil {
		return
	}
	deleteColor := node.Color
	var child *RBNode

	if node.Left == t.nil {
		child = node.Right
		t.transplant(node, node.Right)
	} else if node.Right == t.nil {
		child = node.Left
		t.transplant(node, node.Left)
	} else {
		minNode := t.findMinNode(node.Right)
		deleteColor = minNode.Color
		child = minNode.Right

		if minNode.Parent == node {
			child.Parent = minNode
		} else {
			t.transplant(minNode, minNode.Right)
			minNode.Right = node.Right
			minNode.Right.Parent = minNode
		}
		t.transplant(node, minNode)
		minNode.Left = node.Left
		minNode.Left.Parent = minNode
		minNode.Color = node.Color
	}
	if deleteColor == Black {
		t.deleteFixup(child)
	}
}

// 将 child 置换到node的位置
func (t *RBTree) transplant(node, child *RBNode) {
	if node.Parent == t.nil {
		t.root = child
	} else if node == node.Parent.Left {
		node.Parent.Left = child
	} else {
		node.Parent.Right = child
	}
	child.Parent = node.Parent
}

func (t *RBTree) deleteFixup(node *RBNode) {
	for node != t.root && node.Color == Black {
		if node == node.Parent.Left {
			// node的兄弟节点
			sibling := node.Parent.Right
			// 如果兄弟节点颜色为红色
			if sibling.Color == Red {
				// 将兄弟节点颜色变为黑色
				sibling.Color = Black
				// 将兄弟节点的父节点的颜色变为红色
				node.Parent.Color = Red
				// 左旋
				t.rotateLeft(node.Parent)
				// 将兄弟节点设置为node的兄弟节点
				sibling = node.Parent.Right
			}
			// 如果兄弟节点颜色为黑色
			if sibling.Left.Color == Black && sibling.Right.Color == Black {
				sibling.Color = Red
				node = node.Parent
			} else {

				if sibling.Right.Color == Black {
					sibling.Left.Color = Black
					sibling.Color = Red
					t.rotateRight(sibling)
					sibling = node.Parent.Right
				}
				sibling.Color = node.Parent.Color
				node.Parent.Color = Black
				sibling.Right.Color = Black
				t.rotateLeft(node.Parent)

				node = t.root
			}
		} else {
			//
			sibling := node.Parent.Left
			if sibling.Color == Red {
				sibling.Color = Black
				node.Parent.Color = Red
				t.rotateRight(node.Parent)
				sibling = node.Parent.Left
			}
			if sibling.Right.Color == Black && sibling.Left.Color == Black {
				sibling.Color = Red
				node = node.Parent
			} else {
				if sibling.Left.Color == Black {
					sibling.Right.Color = Black
					sibling.Color = Red
					t.rotateLeft(sibling)
					sibling = node.Parent.Left
				}
				sibling.Color = node.Parent.Color
				node.Parent.Color = Black
				sibling.Left.Color = Black
				t.rotateRight(node.Parent)
				node = t.root
			}
		}

	}
	node.Color = Black

}

func (t *RBTree) findMinNode(node *RBNode) *RBNode {
	if node == t.nil {
		return nil
	}
	for node.Left != t.nil {
		node = node.Left
	}
	return node
}

func (t *RBTree) Print() {
	t.print(t.root, 0)
}
func (t *RBTree) print(node *RBNode, depth int) {
	if node == t.nil {
		return
	}
	t.print(node.Right, depth+1)
	for i := 0; i < depth; i++ {
		fmt.Print("   ")
	}
	if node.Color == Red {
		fmt.Printf("%d(R)\n", node.Value)
	} else {
		fmt.Printf("%d(B)\n", node.Value)
	}
	t.print(node.Left, depth+1)
}

func (t *RBTree) IsRedBlackTree() bool {
	if t.root == nil {
		return true
	}
	return t.isRBTree(t.root)
}

func (t *RBTree) isRBTree(node *RBNode) bool {
	if node == t.nil {
		return true
	}
	if node.Color == Red && (node.Left != nil && node.Left.Color == Red || node.Right != nil && node.Right.Color == Red) {
		return false
	}

	// 4. 递归检查左子树和右子树
	if !t.isRBTree(node.Left) {

		return false
	}
	if !t.isRBTree(node.Right) {

		return false
	}

	leftHeight := t.getBlackHeight(node.Left)
	rightHeight := t.getBlackHeight(node.Right)
	return leftHeight == rightHeight
}

func (t *RBTree) getBlackHeight(node *RBNode) int {
	if node == nil {
		return 0
	}

	if node.Color == Black {
		return 1 + t.getBlackHeight(node.Left)
	}

	return t.getBlackHeight(node.Left)
}

func (t *RBTree) GetRootValue() int {
	if t.root == nil {
		return -1
	}
	return t.root.Value
}

File: typescript/tree/RedBlackTree/rbtree.ts
enum Color {
    Red,
    Black,
}

class RBNode {
    value: number;
    left: RBNode | null;
    right: RBNode | null;
    parent: RBNode | null;
    color: Color;
    constructor(value: number, color: Color = Color.Red) {
        this.value = value;
        this.color = color;
        this.left = this.right = this.parent = null;
    }
}

export class RBTree {
    root: RBNode | null;
    private nil: RBNode = new RBNode(-1, Color.Black);
    constructor() {
        this.root = null;
    }
    insert(value: number) {
        const newNode = new RBNode(value);
        newNode.left = newNode.right = newNode.parent = this.nil;
        this.root = this.insertNode(this.root, newNode, this.nil);
        this.insertFixup(newNode);
    }

    private insertNode(
        node: RBNode | null,
        newNode: RBNode,
        parent: RBNode
    ): RBNode {
        if (node === null || node === this.nil) {
            newNode.parent = parent;
            return newNode;
        }
        if (newNode.value < node.value) {
            node.left = this.insertNode(node.left, newNode, node);
        } else if (newNode.value > node.value) {
            node.right = this.insertNode(node.right, newNode, node);
        }
        return node;
    }

    private insertFixup(node: RBNode) {
        // 当node.parent的颜色为黑色时或者node是根节点时（因为node为根节点是只有可能违反第二条性质，但在代码最后有将根节点标记为黑色）终止循环

        while (node !== this.root && node.parent!.color === Color.Red) {
            // 如果node的父节点是左子节点
            if (node.parent === node.parent!.parent!.left) {
                // 如果node的叔叔节点是红色
                const uncle = node.parent!.parent!.right;
                if (uncle?.color === Color.Red) {
                    // 将node的父节点和叔叔节点的颜色都改为黑色
                    node.parent!.color = Color.Black;
                    uncle.color = Color.Black;
                    // 将node的祖父节点的颜色改为红色
                    node.parent!.parent!.color = Color.Red;
                    // 将node的祖父节点设置为node
                    node = node.parent!.parent!;
                } else {
                    // 如果node的叔叔节点是黑色
                    if (node === node.parent!.right) {
                        node = node.parent!;
                        this.rotateLeft(node);
                    }

                    node.parent!.color = Color.Black;
                    node.parent!.parent!.color = Color.Red;
                    this.rotateRight(node.parent!.parent!);
                }
            } else {
                // 右子树
                const uncle = node.parent!.parent!.left;
                if (uncle?.color === Color.Red) {
                    node.parent!.color = Color.Black;
                    uncle.color = Color.Black;
                    node.parent!.parent!.color = Color.Red;
                    node = node.parent!.parent!;
                } else {
                    if (node === node.parent!.left) {
                        node = node.parent!;
                        this.rotateRight(node);
                    }
                    node.parent!.color = Color.Black;
                    node.parent!.parent!.color = Color.Red;
                    this.rotateLeft(node.parent!.parent!);
                }
            }
        }

        this.root!.color = Color.Black;
    }
    // 左旋转
    private rotateLeft(node: RBNode) {
        const right = node.right;
        node.right = right!.left;
        if (right!.left !== this.nil) {
            right!.left!.parent = node;
        }
        right!.parent = node.parent;
        if (node.parent === this.nil) {
            this.root = right;
        } else if (node === node.parent!.left) {
            node.parent!.left = right;
        } else {
            node.parent!.right = right;
        }
        right!.left = node;
        node.parent = right;
    }
    // 右旋转

    private rotateRight(node: RBNode) {
        const left = node.left;
        node.left = left!.right;
        if (left!.right !== this.nil) {
            left!.right!.parent = node;
        }
        left!.parent = node.parent;
        if (node.parent === this.nil) {
            this.root = left;
        } else if (node === node.parent!.left) {
            node.parent!.left = left;
        } else {
            node.parent!.right = left;
        }
        left!.right = node;
        node.parent = left;
    }

    find(value: number): RBNode | null {
        let node = this.root;
        while (node && node !== this.nil) {
            if (value < node.value) {
                node = node.left;
            } else if (value > node.value) {
                node = node.right;
            } else {
                return node;
            }
        }
        return null;
    }
    findMin(node: RBNode | null): RBNode | null {
        while (node && node !== this.nil) {
            node = node.left;
        }
        return node;
    }

    delete(value: number) {
        const deleteNode = this.find(value);
        if (!deleteNode) return;
        let color = deleteNode.color;
        let child;
        if (deleteNode.left===this.nil) {
            child = deleteNode.right;
            this.transplant(deleteNode, deleteNode.right!);
        } else if (deleteNode.right===this.nil) {
            child = deleteNode.left;
            this.transplant(deleteNode, deleteNode.left!);
        } else {
            const minNode = this.findMin(deleteNode.right)!;
            color = minNode.color;
            child = minNode.right;
            if (minNode.parent === deleteNode) {
                child!.parent = minNode;
            } else {
                this.transplant(minNode, minNode.right);
                minNode.right = deleteNode.right;
                minNode.right!.parent = minNode;
            }
            this.transplant(minNode, minNode.right);
            minNode.right = deleteNode.right;
            minNode.right!.parent = minNode;
            minNode.color = deleteNode.color;
        }

        if (color === Color.Black) {
            this.deleteFixup(child!);
        }
    }
    // 删除之后的rbtree fixup
    deleteFixup(node: RBNode) {
        while (node !== this.root && node.color === Color.Black) {
            if (node === node.parent!.left) {
                let brother = node.parent!.right!;
                if (brother.color === Color.Red) {
                    brother.color = Color.Black;
                    node.parent!.color = Color.Red;
                    this.rotateLeft(node.parent!);
                    brother = node.parent!.right!;
                }
                if (
                    brother.left!.color === Color.Black &&
                    brother.right!.color === Color.Black
                ) {
                    brother.color = Color.Red;
                    node = node.parent!;
                } else {
                    if (brother.right!.color === Color.Black) {
                        brother.left!.color = Color.Black;
                        brother.color = Color.Red;
                        this.rotateRight(brother);
                        brother = node.parent!.right!;
                    }
                    brother.color = node.parent!.color;
                    node.parent!.color = Color.Black;
                    brother.right!.color = Color.Black;

                    this.rotateLeft(node.parent!);
                    node = this.root!;
                }
            } else {
                let brother = node.parent!.left!;
                if (brother.color === Color.Red) {
                    brother.color = Color.Black;
                    node.parent!.color = Color.Red;
                    this.rotateRight(node.parent!);
                    brother = node.parent!.left!;
                }
                if (
                    brother.left!.color === Color.Black &&
                    brother.right!.color === Color.Black
                ) {
                    brother.color = Color.Red;
                    node = node.parent!;
                } else {
                    if (brother.left!.color === Color.Black) {
                        brother.right!.color = Color.Black;
                        brother.color = Color.Red;
                        this.rotateLeft(brother);
                        brother = node.parent!.left!;
                    }
                    brother.color = node.parent!.color;
                    node.parent!.color = Color.Black;
                    brother.left!.color = Color.Black;

                    this.rotateRight(node.parent!);
                    node = this.root!;
                }
            }
        }
        node.color = Color.Black;
    }
    // 将child 置换到 node 的位置
    transplant(node: RBNode, child: RBNode | null) {
        if (node.parent === this.nil) {
            this.root = child;
        } else if (node === node.parent!.left) {
            node.parent!.left = child;
        } else {
            node.parent!.right = child;
        }
        child!.parent = node.parent;
    }

    private isRed(node: RBNode | null): boolean {
        return node?.color === Color.Red;
    }
    private isBlack(node: RBNode | null): boolean {
        return node?.color === Color.Black;
    }
    private isNil(node: RBNode | null): boolean {
        return node === this.nil;
    }

    getBlackHeight(node: RBNode): number {
        if (!node)
            return 0; // 空节点黑色高度为0
        if (node.color == Color.Black)
            return 1 + this.getBlackHeight(node.left!); // 黑色节点加1，并递归左子树
        return this.getBlackHeight(node.left!);         // 红色节点不增加高度，直接递归左子树
    }
    /**
     * 检查这棵树是否符合红黑树的性质
     * @returns 
     */
    isRedBlackTree(): boolean {
        if (!this.root)
            return true

        // 根节点必须是黑色
        if (this.isRed(this.root))
            return false;

        return this.isRedBlackTreeHelper(this.root);
    }
    private isRedBlackTreeHelper(node: RBNode): boolean {
        if (!node)
            return true; // 空节点视为满足条件

        // 3. 检查红色节点的孩子是否为黑色
        if (this.isRed(node)) {
            if (this.isRed(node.left) || this.isRed(node.right)) {
                return false;
            }
        }

        // 4. 递归检查左子树和右子树
        if (!this.isRedBlackTreeHelper(node.left!))
            return false;
        if (!this.isRedBlackTreeHelper(node.right!))
            return false;

        // 5. 检查从当前节点到其所有叶子节点的简单路径上黑色节点的数量是否相同
        // 这通常需要一个额外的函数来计算黑色高度，并比较左右子树的黑色高度
        const leftBlackHeight = this.getBlackHeight(node.left!);
        const rightBlackHeight = this.getBlackHeight(node.right!);
        if (leftBlackHeight != rightBlackHeight)
            return false;

        // 所有检查都通过，返回true
        return true;
    }
}
